[REEF-910] Assemble entire driver configuration on driver instead of client

afchung opened the pull request:
This addressed the issue by
  * Adds the Bootstrap Launcher for YARN, which removes Java dependency if submitted through REST.
  * Modifies DriverLauncher to accomodate other types of launchers.
  * Moves YARN submission through .NET into the new two-step process.

JIRA:
  [REEF-910](https://issues.apache.org/jira/browse/REEF-910)

jwang98052 commented:
I will review it.

jwang98052 commented on lang/java/reef-bridge-client/src/main/avro/JobSubmissionParameters.avsc:
```diff
@@ -49,6 +49,7 @@
         { "name": "sharedJobSubmissionParameters", "type": "AvroJobSubmissionParameters" },
         { "name": "driverMemory", "type": "int" },
         { "name": "driverRecoveryTimeout", "type": "int" },
+		{ "name": "dfsJobSubmissionFolder", "type": ["null", "string"], "default" : null },
```
indent

afchung replied:
Done.

jwang98052 commented on lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/launch/JavaLaunchCommandBuilder.java:
```diff
@@ -168,6 +180,28 @@ public JavaLaunchCommandBuilder addOption(final String option) {
+   * @param copyNull  create an empty parameter if the property is missing in current process.
+   * @param propNames property names.
+   */
+  private static void propagateProperties(
```
not necessarily to be static

afchung replied:
why not?

jwang98052 replied:
It is only used inside the class even it doesn't access any instance variables. Agree it is just a preference. This one is more like a utility method.  I am fine with it.

jwang98052 commented on lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java:
```diff
@@ -201,10 +202,33 @@ public YarnSubmissionHelper setQueue(final String queueName) {
     // SET EXEC COMMAND
-    final List<String> launchCommand = new JavaLaunchCommandBuilder(commandPrefixList)
-        .setConfigurationFileName(this.fileNames.getDriverConfigurationPath())
+    final List<String> launchCommand = new JavaLaunchCommandBuilder(launcherClazz, commandPrefixList)
```
Why launcherClazz is passed through the constructor but others are set by set methods?

afchung replied:
Because it's a required parameter, and I didn't want this particular required parameter to be "hidden" with {{set}}. The parameter should only be set once anyway.

anupam128 commented on lang/cs/Org.Apache.REEF.Client/Avro/YARN/AvroYarnJobSubmissionParameters.cs:
```diff
@@ -28,7 +28,7 @@ namespace Org.Apache.REEF.Client.Avro.YARN
     public sealed class AvroYarnJobSubmissionParameters
     {
-        private const string JsonSchema = @"{""type"":""record"",""name"":""org.apache.reef.reef.bridge.client.avro.AvroYarnJobSubmissionParameters"",""doc"":""General cross-language submission parameters to the YARN runtime"",""fields"":[{""name"":""sharedJobSubmissionParameters"",""type"":{""type"":""record"",""name"":""org.apache.reef.reef.bridge.client.avro.AvroJobSubmissionParameters"",""doc ...
+        private const string JsonSchema = @"{""type"":""record"",""name"":""org.apache.reef.reef.bridge.client.avro.AvroYarnJobSubmissionParameters"",""doc"":""General cross-language submission parameters to the YARN runtime"",""fields"":[{""name"":""sharedJobSubmissionParameters"",""type"":{""type"":""record"",""name"":""org.apache.reef.reef.bridge.client.avro.AvroJobSubmissionParameters"",""doc ...
```
Nit: is it possible to make more readable?

anupam128 commented:
In general changes LGTM. I would suggest you to add unittests for appropriate logic.

jwang98052 commented on lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnJobSubmissionClient.java:
```diff
@@ -131,29 +121,26 @@ private void launch(final YarnSubmissionFromCS yarnSubmission) throws IOExceptio
+          .setPriority(yarnSubmission.getPriority())
+          .setQueue(yarnSubmission.getQueue())
+          .setMaxApplicationAttempts(yarnSubmission.getMaxApplicationSubmissions())
+          .setPreserveEvaluators(yarnSubmission.getDriverRecoveryTimeout() > 0)
```
Seems very different from original one.

jwang98052 commented on lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/launch/JavaLaunchCommandBuilder.java:
```diff
@@ -48,21 +48,33 @@
   private Boolean assertionsEnabled = null;
   private Map<String, JVMOption> options = new HashMap<>();
```
final

afchung replied:
Done

jwang98052 commented:
I am done my pass.

afchung commented:
@jwang98052 Addressed your comments, please have another look, thanks! Also, if it's not too much trouble, could you test the C# portion on HDInsight as well? They are passing on my side.

anupam128 commented:
+1

jwang98052 commented on lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnBootstrapDriverConfigGenerator.java:
```diff
@@ -0,0 +1,143 @@
+ * the Driver configuration at runtime. Called by {@link YarnBootstrapREEFLauncher}.
+ */
+final class YarnBootstrapDriverConfigGenerator {
+  private static final REEFFileNames REEF_FILE_NAMES = new REEFFileNames();
```
Should be injected in the constructor instead of new.

jwang98052 commented on lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnBootstrapDriverConfigGenerator.java:
```diff
@@ -0,0 +1,143 @@
+    this.configurationSerializer = configurationSerializer;
+  }
+
+  public String writeDriverConfigurationFile(final String bootstrapArgsLocation) throws IOException {
```
can be internal?

afchung replied:
The class is "internal" anyway.
